zcash_primitives/src/transaction/builder.rs

```diff
@@ -302,18 +306,20 @@ impl TransactionMetadata {
 }
 
 /// Generates a [`Transaction`] from its inputs and outputs.
-pub struct Builder<R: RngCore + CryptoRng> {
+pub struct Builder<P: consensus::Parameters, R: RngCore + CryptoRng> {
     rng: R,
+    height: u32,
     mtx: TransactionData,
     fee: Amount,
     anchor: Option<Fr>,
     spends: Vec<SpendDescriptionInfo>,
     outputs: Vec<SaplingOutput>,
     transparent_inputs: TransparentInputs,
     change_address: Option<(OutgoingViewingKey, PaymentAddress<Bls12>)>,
+    phantom: PhantomData<P>,
 }
 
-impl Builder<OsRng> {
+impl Builder<MainNetwork, OsRng> {
```

Comment:
i added separate `impl Builder<MainNetwork, OsRng>` and `impl Builder<TestNetwork, OsRng>` because the `impl` would not take an associated type binding (i.e. `impl Builder<P: consensus::Parameters, OsRng>`)

Reply:
The way you would write this is `impl<P: consensus::Parameters> Builder<P, OsRng>`.